sections/Chapter 7 Sprint 5/virtuoso_part_2.tex

```diff
@@ -0,0 +1,115 @@
+
+\begin{lstlisting}[language=CSharp, caption={Data insertion endpoint and data insertion logic}, label={lst:virtuoso_data_insertion}]
+[HttpPost, Route("/[controller]/")]
+public async Task<IActionResult> Insert(string turtle, string? graph)
+{
+    string? virtuosoEndpoint = Environment.GetEnvironmentVariable("VIRTUOSO_ENDPOINT");
+    if (virtuosoEndpoint == null)
+    {
+        return StatusCode((int)HttpStatusCode.InternalServerError, "VIRTUOSO_ENDPOINT environment variable not set");
+    }
+
+    try
+    {
+        string insertResponse = !string.IsNullOrEmpty(graph) ?
+            await new VirtuosoDataStore(virtuosoEndpoint).InsertTurtleGraph(turtle, graph) :
+            await new VirtuosoDataStore(virtuosoEndpoint).InsertTurtleGraph(turtle);
+
+        return Ok(insertResponse);
+    }
+    catch (Exception e)
+    {
+        return StatusCode(503, e.Message);
+    }
+}
+
+public async Task<string> InsertTurtleGraph(string turtle, string? graphName = "knox")
+{
+    IGraph g = new Graph()
+    {
+        BaseUri = uri,
+    };
+    
+    GraphHandler graphHandler = new(g);
+    TurtleParser parser = new();
+    parser.Load(graphHandler, new StringReader(turtle));
+    
+    Console.WriteLine($"Triples in given graph: {g.Triples.Count}");
+
+    string[] chunkedQueries = ChunkRecords(turtle, "\n\n", 50);
+
+    foreach (string chunkedQuery in chunkedQueries)
+    {
+        StringBuilder sb = new();
+        
+        sb.Append("INSERT DATA { GRAPH <" + graphName + "> {");
+        sb.Append(chunkedQuery);
+        sb.Append("} }");
+
+        await Query(sb.ToString());
+    }
+
+    return "OK";
+}
+\end{lstlisting}
+
+
+We have also implemented chunking logic for data insertion, such that users do not get an error when sending a large amount of data. This was done because we, during our tests, experienced that Virtuoso would error out when sending a large amount of data --- it could not execute such large amounts of generated SQL code. The logic for diving chunks of data into smaller chunks is implemented in the \texttt{ChunkRecords} function, which can be seen in code snippet \ref{lst:virtuoso_chunking}. The logic itself is to divide a large string of input data into smaller chunks given some separator and chunk size. This is done by first splitting the string into an array of strings with the separator, and then combining individual strings into chunks with the given size.
```

Comment:
```suggestion
We have also implemented chunking logic for data insertion, such that users do not get an error when sending a large amount of data. This was done because we, during our tests, experienced that Virtuoso would throw an exception when sending a large amount of data --- it could not execute such large amounts of generated SQL code. The logic for dividing chunks of data into smaller chunks is implemented in the \texttt{ChunkRecords} function, which can be seen in code snippet \ref{lst:virtuoso_chunking}. This is done by dividing a large string of input data into smaller chunks, given some separator and chunk size. This is done by first splitting the string into an array of strings with the separator, and then combining individual strings into chunks with the given size.
```